crates/pop-common/src/metadata.rs

```diff
@@ -0,0 +1,226 @@
+// SPDX-License-Identifier: GPL-3.0
+
+use scale_info::{form::PortableForm, PortableRegistry, Type, TypeDef, TypeDefPrimitive};
+
+/// Formats a specified type, using the registry to output its full type representation.
+///
+/// # Arguments
+/// * `ty`: A reference to the `Type<PortableForm>` to be formatted.
+/// * `registry`: A reference to the `PortableRegistry` to resolve type dependencies.
+pub fn format_type(ty: &Type<PortableForm>, registry: &PortableRegistry) -> String {
+	let mut name = ty
+		.path
+		.segments
+		.last()
+		.map(|s| s.to_owned())
+		.unwrap_or_else(|| ty.path.to_string());
+
+	if !ty.type_params.is_empty() {
+		let params: Vec<_> = ty
+			.type_params
+			.iter()
+			.filter_map(|p| {
+				if let Some(ty) = p.ty {
+					registry.resolve(ty.id)
+				} else {
+					None // Ignore if p.ty is None
+				}
+			})
+			.map(|t| format_type(t, registry))
+			.collect();
+		name = format!("{name}<{}>", params.join(","));
+	}
+	name = format!(
+		"{name}{}",
+		match &ty.type_def {
+			TypeDef::Composite(composite) => {
+				if composite.fields.is_empty() {
+					return "".to_string();
+				}
+
+				let mut named = false;
+				let fields: Vec<_> = composite
+					.fields
+					.iter()
+					.filter_map(|f| match f.name.as_ref() {
+						None => registry.resolve(f.ty.id).map(|t| format_type(t, registry)),
+						Some(field) => {
+							named = true;
+							f.type_name.as_ref().map(|t| format!("{field}: {t}"))
+						},
+					})
+					.collect();
+				match named {
+					true => format!(" {{ {} }}", fields.join(", ")),
+					false => format!(" ({})", fields.join(", ")),
+				}
+			},
+			TypeDef::Variant(variant) => {
+				let variants: Vec<_> = variant
+					.variants
+					.iter()
+					.map(|v| {
+						if v.fields.is_empty() {
+							return v.name.clone();
+						}
+
+						let name = v.name.as_str();
+						let mut named = false;
+						let fields: Vec<_> = v
+							.fields
+							.iter()
+							.filter_map(|f| match f.name.as_ref() {
+								None => registry.resolve(f.ty.id).map(|t| format_type(t, registry)),
+								Some(field) => {
+									named = true;
+									f.type_name.as_ref().map(|t| format!("{field}: {t}"))
+								},
+							})
+							.collect();
+						format!(
+							"{name}{}",
+							match named {
+								true => format!("{{ {} }}", fields.join(", ")),
+								false => format!("({})", fields.join(", ")),
+							}
+						)
+					})
+					.collect();
+				format!(": {}", variants.join(", "))
+			},
+			TypeDef::Sequence(sequence) => {
+				format!(
+					"[{}]",
+					format_type(
+						registry.resolve(sequence.type_param.id).expect("sequence type not found"),
```

Comment:
I see some expects in this code. Not sure what the practise is in pop cli but figure we prefer to prevent using them?

Reply:
As mentioned in [this comment](https://github.com/r0gue-io/pop-cli/pull/316#discussion_r1867800979), this is not a new function added in this PR; it was simply moved from the `pop_contract` crate to the `pop_common` crate. I believe it would be better to address any related changes or improvements in a separate PR.